agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -57,137 +97,242 @@ class PythonRefAgentConnection : AgentConnection {
+                         */
+                        sendRequest(pubKey)
+                        /**
+                         * The agent receives the response from the other party and informs the client with "response_received"
+                         * message
+                         */
+                        webSocket.receiveMessageOfType<ObjectNode>(MESSAGE_TYPES.RESPONSE_RECEIVED, pubKey).subscribe({
+                            /**
+                             * The "response_received" message will contain the other party's DID and endpoint
+                             */
+                            val theirPubkey = it["history"]["connection"]["DIDDoc"]["publicKey"][0]["publicKeyBase58"].asText()
+                            val theirDid = it["their_did"].asText()
+                            /**
+                             * retrieve my_did from agent's STATE
+                             */
+                            sendAsJson(StateRequest())
+                            webSocket.receiveMessageOfType<ObjectNode>(MESSAGE_TYPES.STATE_RESPONSE).subscribe { stateResponse ->
+                                val pairwise = stateResponse["content"]["pairwise_connections"].find { node ->
+                                    node["metadata"]["their_vk"].asText() == theirPubkey
+                                }
+                                if (pairwise != null) {
+                                    val indyParty = IndyParty(webSocket, theirDid, invRcv.endpoint, theirPubkey, pairwise["my_did"].asText())
+                                    indyParties[theirDid] = indyParty
+                                    observer.onSuccess(indyParty)
+                                }
+                            }
+                        }, { e: Throwable -> throw(e) })
+                    }, { e: Throwable -> throw(e) })
+                } else {
+                    throw AgentConnectionException("AgentConnection object has wrong state")
+                }
+            } catch (e: Throwable) {
+                observer.onError(e)
+            }
         }
     }
 
-    fun sendJson(obj: Any) = webSocket.sendJson(obj)
-
-    override fun generateInvite(): String {
-        webSocket.sendJson(SendMessage(type = MESSAGE_TYPES.GENERATE_INVITE, id = Instant.now().toEpochMilli().toString()))
-        return webSocket.waitForMessageOfType<ReceiveInviteMessage>(MESSAGE_TYPES.INVITE_GENERATED).invite
-    }
-
-    override fun waitForInvitedParty() {
-        if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTED) {
-            webSocket.waitForMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED).also {
-                counterParty = IndyParty(it.did, it.endpoint)
-                sendJson(RequestSendResponseMessage(it.did))
+    private fun sendAsJson(obj: Any) = webSocket.sendAsJson(obj)
+
+    /**
+     * Retrieves a new invite from the agent
+     */
+    override fun generateInvite(): Single<String> {
```

Comment:
I would suggest to move Invite acceptance from waitForInvitedParty to generateInvite, because the logic is very connected

```
private fun subscribeOnInviteAccept(timeout: Long) {
        webSocket.receiveMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED)
                .timeout(timeout, TimeUnit.MILLISECONDS)
                .subscribe({
                    log.info { "Request to accept invite for ${it.did} has been received" }
                    sendAsJson(RequestSendResponseMessage(it.did))

                    webSocket.receiveMessageOfType<RequestResponseSentMessage>(MESSAGE_TYPES.RESPONSE_SENT)
                            .subscribe { responseSent -> log.info { "Invite for ${it.did} has been accepted" } }
                })
    }

    /**
     * Retrieves a new invite from the agent
     */
    override fun generateAndAcceptInvite(timeout: Long = 60000L): Single<String> {
        return Single.create { observer ->
            try {
                /**
                 * to generate the invite, send "generate_invite" message to the agent, and wait for "invite_generated"
                 * which must contain an invite
                 */
                webSocket.sendAsJson(SendMessage(type = MESSAGE_TYPES.GENERATE_INVITE))
                webSocket.receiveMessageOfType<ReceiveInviteMessage>(MESSAGE_TYPES.INVITE_GENERATED).subscribe({ msg ->
                    subscribeOnInviteAccept(timeout)
                    observer.onSuccess(msg.invite)
                }, { e: Throwable -> throw(e) })
            } catch (e: Throwable) {
                observer.onError(e)
            }
        }
    }
```

Reply:
There's a problem here, that generateAndAcceptInvite() produces an invite that can only be accepted while the node runs. Once the node has been restarted, it cannot waitForInvitedParty() anymore with the same invite, because REQUEST_RECEIVED message will probably remain unanswered. The approach when the invite is generated and accepted separately is more reliable.

Reply:
and flexible.

Reply:
It's fine. In this case a restore flow should be run. 
Please, fix invalid sequence of messages.

Reply:
Sure, I will. I still prefer to subscribeOnInviteAccept() from the context of waitForInvitedParty() and without the timeout. We must be able to wait for invited party at any time, because mobile users may have arbitrary long timeouts between receiving and accepting invites.

Reply:
done in 9cf8faa08e968dad63c240f8d248f5cbdcbbfac7